frontend/rust-lib/flowy-document/src/event_handler.rs

```diff
@@ -401,3 +406,125 @@ pub(crate) async fn convert_data_to_json_handler(
 
   data_result_ok(ConvertDataToJsonResponsePB { json: result })
 }
+
+// Handler for uploading a file
+// `workspace_id` and `file_name` determines file identity
+pub(crate) async fn upload_file_handler(
+  params: AFPluginData<UploadFileParamsPB>,
+  manager: AFPluginState<Weak<DocumentManager>>,
+) -> DataResult<UploadedFilePB, FlowyError> {
+  let manager = upgrade_document(manager)?;
+  let file_service = manager
+    .get_file_storage_service()
+    .upgrade()
+    .ok_or(FlowyError::internal().with_context("The file storage service is already dropped"))?;
+
+  let mut file = tokio::fs::File::open(&params.local_file_path).await?;
+  let mut content = Vec::new();
+  let n = file.read_to_end(&mut content).await?;
+  info!("read {} bytes from file: {}", n, &params.local_file_path);
+
+  let hash = fxhash::hash(&content);
+  let mime = mime_guess::from_path(&params.local_file_path).first_or_octet_stream();
+
+  let url = file_service
+    .get_object_url(ObjectIdentity {
+      workspace_id: params.workspace_id.to_owned(),
+      file_id: hash.to_string(),
+    })
+    .await?;
+
+  // let the upload happen in the background
+  let url_clone = url.clone();
+  tokio::spawn(async move {
+    if let Err(e) = file_service
+      .put_object(
+        url_clone,
+        ObjectValue {
```

Comment:
The suggestion here is to encapsulate the logic for creating an `ObjectValue` from a file into a dedicated method (`from_file`) within the `ObjectValue` struct or associated implementation in Rust. This approach offers several benefits in terms of code organization and dependency management. Let me clarify this:

**Suggested Refactoring**:
- Create a `from_file` method within the `ObjectValue` implementation.
- This method will handle the process of opening a file, reading its contents, and determining the necessary metadata (like hash and mime type).

**New Function in `ObjectValue`**:

```rust
impl ObjectValue {
    async fn from_file(local_file_path: &str) -> Result<Self, io::Error> {
        let mut file = tokio::fs::File::open(local_file_path).await?;
        let mut content = Vec::new();
        let n = file.read_to_end(&mut content).await?;
        info!("read {} bytes from file: {}", n, local_file_path);

        let hash = fxhash::hash(&content);
        let mime = mime_guess::from_path(local_file_path).first_or_octet_stream();

        // Construct and return the ObjectValue based on the file content and metadata
        Ok(ObjectValue {
            // ... fields populated with content, hash, mime, etc.
        })
    }
}
```

**Benefits**:

1. **Code Reusability**: By centralizing the file reading and ObjectValue construction logic in `from_file`, you avoid duplicating this code wherever you need to create an `ObjectValue` from a file. This makes the codebase DRY (Don't Repeat Yourself), easier to maintain, and less error-prone.

2. **Reduced Dependency Spread**: The dependencies `fxhash`, `mime`, and `mime_guess` are only needed within the `ObjectValue` implementation. Other parts of the codebase that create `ObjectValue` instances don't need to directly depend on these libraries, reducing the spread and management of dependencies.

3. **Simplified Interface**: External code that needs to create an `ObjectValue` from a file can do so with a single, straightforward method call, without needing to understand the underlying details of file handling and metadata processing.

Reply:
moved logic